src/js_types.hpp

```diff
@@ -351,13 +351,28 @@ REALM_JS_INLINE void set_internal(const typename T::Object &object, typename Cla
 
 template<typename T>
 inline bool Value<T>::is_valid_for_property(ContextType context, const ValueType &value, const Property& prop)
-{
-    if (prop.is_nullable && (is_null(context, value) || is_undefined(context, value))) {
+{ 
+    if (is_nullable(prop.type) && (is_null(context, value) || is_undefined(context, value))) {
         return true;
     }
 
+    if (realm::is_array(prop.type)) {
+        if (prop.type == PropertyType::Object) {
+            auto object = to_object(context, value);
+            return Object<T>::template is_instance<ResultsClass<T>>(context, object)
+                || Object<T>::template is_instance<ListClass<T>>(context, object);
+        } else if (prop.type == PropertyType::LinkingObjects) {
+            return false;
+        }
+        else if (prop.type == PropertyType::Array) {
```

Comment:
I don't think this check makes sense. `is_array` checks if the `PropertyType::Array` is set, and if it is you then test against the various types to determine what the array contains (objects, linking objects, and in the future the various primitive types).